pwiz_tools/Shared/PanoramaClient/PanoramaUtil.cs

```diff
@@ -675,37 +349,128 @@ private static string GetErrorMessage(FolderState state, string folderPath, stri
+        public ErrorMessageBuilder Exception(Exception ex)
         {
-            var message = mainMessage;
+            Exception(ex, null);
+            return this;
+        }
 
-            if (error != null || uri != null)
+        // If the given exception is a WebException, this method tries to parse the response as JSON and
+        // get a LabKey error if it was set. The method accepts a function to parse the response in the 
+        // WebException so that tests can pass in a custom function.
+        public ErrorMessageBuilder Exception(Exception ex, Func<WebException, LabKeyError> labkeyErrorGetter)
+        {
+            _exceptionMessage = ex.Message;
+            var exception = ex as WebException;
+            if (exception != null && labkeyErrorGetter != null)
             {
-                var sb = new StringBuilder();
+                _labkeyError = labkeyErrorGetter(exception);
+            }
 
-                if (error != null)
-                {
-                    sb.AppendLine(string.Format(Resources.GenericState_AppendErrorAndUri_Error___0_, error));
-                }
+            return this;
+        }
 
-                if (uri != null)
-                {
-                    sb.AppendLine(string.Format(Resources.GenericState_AppendErrorAndUri_URL___0_, uri));
-                }
+        public ErrorMessageBuilder Uri(Uri requestUri)
+        {
+            _uri = requestUri;
+            return this;
+        }
 
-                message = TextUtil.LineSeparate(message, string.Empty, sb.ToString());
+        public ErrorMessageBuilder Response(JObject json)
+        {
+            if (json != null)
+            {
+                _responseString = json.ToString(Formatting.Indented);
             }
+            return this;
+        }
+        public ErrorMessageBuilder Response(string response)
+        {
+            _responseString = response;
+            return this;
+        }
 
-            return message;
+        public string Build()
+        {
+            var sb = new StringBuilder();
+            if (!string.IsNullOrEmpty(_errorDetail) || !string.IsNullOrEmpty(_exceptionMessage) || _labkeyError != null)
+            {
+                sb.Append(Resources.ErrorMessageBuilder_Build_Error__);
+                if (!string.IsNullOrEmpty(_errorDetail)) sb.AppendLine(_errorDetail);
+                if (!string.IsNullOrEmpty(_exceptionMessage)) sb.AppendLine(_exceptionMessage);
+                if (_labkeyError != null) sb.AppendLine(_labkeyError.ToString());
+            }
+            if (_uri != null)
+            {
+                sb.AppendLine(string.Format(Resources.GenericState_AppendErrorAndUri_URL___0_, _uri.AbsoluteUri));
+            }
+
+            if (!string.IsNullOrEmpty(_responseString))
+            {
+                // TODO: Should we truncate the response string, or display it at all?
+                //       CommonAlertDlg truncates the displayed string to 50000. 

```

Comment:
FWIW, CommonAlertDlg hangs onto the entire message in case the user presses the copy button. It does truncate the message to display it.